src/Sylius/Behat/spec/Context/Domain/ProductContextSpec.php

```diff
@@ -0,0 +1,124 @@
+ *
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class ProductContextSpec extends ObjectBehavior
+{
+    function let(ProductVariantRepositoryInterface $productVariantRepository, SharedStorageInterface $sharedStorage)
+    {
+        $this->beConstructedWith($sharedStorage, $productVariantRepository);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Behat\Context\Domain\ProductContext');
+    }
+
+    function it_implements_context_interface()
+    {
+        $this->shouldImplement(Context::class);
+    }
+
+    function it_deletes_product_variant_from_repository(
+        ProductVariantRepositoryInterface $productVariantRepository,
+        ProductVariantInterface $productVariant
+    ) {
+        $productVariantRepository->remove($productVariant)->shouldBeCalled();
+
+        $this->iDeleteTheVariantOfProduct($productVariant);
+    }
+
+    function it_tries_to_delete_product_variant_from_repository_which_should_not_be_deleted(
+        ProductVariantRepositoryInterface $productVariantRepository,
+        SharedStorageInterface $sharedStorage,
+        ProductVariantInterface $productVariant
+    ) {
+        $productVariantRepository->remove($productVariant)->willThrow(DBALException::class);
+        $sharedStorage->set('exception', Argument::type(DBALException::class))->shouldBeCalled();
+
+        $this->iTryToDeleteTheVariantOfProduct($productVariant);
+    }
+
+    function it_checks_if_a_product_variant_exists_in_repository(
+        ProductVariantRepositoryInterface $productVariantRepository,
+        ProductVariantInterface $productVariant
+    ) {
+        $productVariant->getId()->willReturn(1);
+        $productVariantRepository->find(1)->willReturn(null);
+
+        $this->productVariantShouldNotExistInTheProductCatalog($productVariant);
+    }
+
+    function it_throws_an_exception_if_a_product_variant_still_exists(
+        ProductVariantRepositoryInterface $productVariantRepository,
+        ProductVariantInterface $productVariant
+    ) {
+        $productVariant->getId()->willReturn(1);
+        $productVariantRepository->find(1)->willReturn($productVariant);
+
+        $this->shouldThrow(NotEqualException::class)->during('productVariantShouldNotExistInTheProductCatalog', [$productVariant]);
+    }
+
+    function it_checks_if_proper_exception_has_been_thrown_in_a_previous_step(
+        SharedStorageInterface $sharedStorage,
+        DBALException $exception
+    ) {
+        $sharedStorage->get('exception')->willReturn($exception);
+
+        $this->iShouldBeNotifiedThatThisVariantIsInUseAndCannotBeDeleted();
+    }
+
+    function it_throws_an_exception_if_a_wrong_exception_was_passed(
+        SharedStorageInterface $sharedStorage,
+        \Exception $exception
+    ) {
+        $sharedStorage->get('exception')->willReturn($exception);
+
+        $this->shouldThrow(FailureException::class)->during('iShouldBeNotifiedThatThisVariantIsInUseAndCannotBeDeleted', []);
```

Comment:
Missing scenario when there will be no exception. `SharedStorage` will throw `\InvalidArgumentException` maybe just change typehint on this spec?